### PR TITLE
Add safe guard for provisioning/terminating TPU VM and fix spot launch TPU resource leak 

### PR DESCRIPTION
**Pull request by infwinston**
Our user reported some errors when getting TPU IPs. This PR adds some safe guard for `_get_tpu_vm_pod_ips` with better error handling to prevent failures like leaked resources with duplicated cluster IDs.

This PR also fixed https://github.com/skypilot-org/skypilot/issues/1514 which tried to remove an `INIT` TPU VM resource.

Tested:
- [x] launching TPU VM/TPU Pod
- [x] TPU Pod smoke test

### CODE REVIEW

File: sky/spot/recovery_strategy.py [hunk 1]
```diff
@@ -305,6 +306,13 @@ def recover(self) -> float:
                 new_resources = resources.copy(cloud=launched_cloud,
                                                region=launched_region)
                 task.set_resources({new_resources})
+
+                # Clean up preempted TPU VM before launching the cluster.
+                # This is needed as status -r will not remove it if GCP
+                # turns the VM state to other than PREEMPTED.
+                is_tpuvm = tpu_utils.is_tpu_vm(new_resources)
+                if is_tpuvm:
+                    self.terminate_cluster()
```
Comment (Michaelvll) [inline sky/spot/recovery_strategy.py]: After discussing offline, we found it might be better to have this special case handling in the controller process instead of the recovery strategy. Therefore, future recovery strategies will not need to handle separately.

**Comment by infwinston** [comment]
Ok after some investigation, I finally realize what's going on. So basically when a TPU VM is preempted by GCP (i.e., no longer SSHable into the server), its state may not be immediately set to `PREEMPTED` but may stay in `READY` for a while.
Hence, even when our controller detects such failure to connect and determine the server is preempted, `status -r` won't reflect such status correctly.


During `status -r`, we run the below code to refresh status
https://github.com/skypilot-org/skypilot/blob/742c1dc82e62322d4e6d61c68b021b816f6110ad/sky/backends/backend_utils.py#L1655
and then we got `READY` which translated to `ClusterStatus.UP` (see below log)

```
(sky-de99-weichiang, pid=734143) I 12-10 21:39:36 spot_utils.py:72] Failed to connect to the cluster.
(sky-de99-weichiang, pid=734143) I 12-10 21:39:36 spot_utils.py:73] ==================================
(sky-de99-weichiang, pid=734143) D 12-10 21:39:38 backend_utils.py:1676] Refreshing status: Failed to get IPs from cluster 'sky-de99-weichiang-16', trying to fetch from provider.
(sky-de99-weichiang, pid=734143) D 12-10 21:39:39 backend_utils.py:1433] gcloud compute tpus tpu-vm list --zone us-central1-f --filter="(labels.ray-cluster-name=sky-de99-weichiang-16 AND labels.ray-launch-config=(c070152c2602f36dd1adc9a0a6cd087fc2db9352))" --format="value(state)" returned 0.
(sky-de99-weichiang, pid=734143) D 12-10 21:39:39 backend_utils.py:1433] **** STDOUT ****
(sky-de99-weichiang, pid=734143) D 12-10 21:39:39 backend_utils.py:1433] READY
(sky-de99-weichiang, pid=734143) D 12-10 21:39:39 backend_utils.py:1433]
(sky-de99-weichiang, pid=734143) D 12-10 21:39:39 backend_utils.py:1433] **** STDERR ****
(sky-de99-weichiang, pid=734143) D 12-10 21:39:39 backend_utils.py:1433]
(sky-de99-weichiang, pid=734143) W 12-10 21:39:39 backend_utils.py:1452] Cluster status: 'READY'.
(sky-de99-weichiang, pid=734143) D 12-10 21:39:39 backend_utils.py:1710] Failed to reset autostop. Due to <class 'sky.exceptions.CommandError'>: Command python3 -u -c 'from sky.skylet import autostop_lib;autostop_lib.set_autostop(-1, '"'"'cloudvmray'"'"', False)' failed with return code 255.
(sky-de99-weichiang, pid=734143) D 12-10 21:39:39 backend_utils.py:1710] Failed to set autostop
(sky-de99-weichiang, pid=734143) I 12-10 21:39:39 controller.py:118] Cluster is preempted (status: INIT). Recovering...
(sky-de99-weichiang, pid=734143) I 12-10 21:42:11 spot_state.py:134] === Recovering... ===
(sky-de99-weichiang, pid=734143) I 12-10 21:42:11 recovery_strategy.py:126] Ignoring the job cancellation failure; the spot cluster is likely completely stopped.
(sky-de99-weichiang, pid=734143) I 12-10 21:42:11 recovery_strategy.py:126]   Detailed exception: 'NoneType' object has no attribute 'cluster_yaml'
(sky-de99-weichiang, pid=734143) D 12-10 21:42:11 recovery_strategy.py:316] Terminating unhealthy spot cluster.
(sky-de99-weichiang, pid=734143) D 12-10 21:42:11 recovery_strategy.py:320] Relaunch the cluster  without constraining to prior cloud/region.
```

However, later we immediately change the status from `UP` to `INIT` because we determine the cluster is "abnormal".
https://github.com/skypilot-org/skypilot/blob/742c1dc82e62322d4e6d61c68b021b816f6110ad/sky/backends/backend_utils.py#L1674

https://github.com/skypilot-org/skypilot/blob/742c1dc82e62322d4e6d61c68b021b816f6110ad/sky/backends/backend_utils.py#L1696


That's why we see the cluster was in `INIT` state.
```
(sky-de99-weichiang, pid=734143) I 12-10 21:39:39 controller.py:118] Cluster is preempted (status: INIT). Recovering...
```

===================

In conclusion
1) We cannot rely on `status -r` to clean up preempted resources because the state may not reflect in real-time. 
2) We also cannot rely on `ray up` because a spot controller may not try `ray up` with the same region/zone after preemption. so a preempted resource can still be leaked.
in the end, I think it is the spot controller's job to make sure there's no leaked preempted resource before launching a new one.

To avoid adding specific logic that only applies to TPU VM, I introduce a slightly better abstraction in this PR.
https://github.com/skypilot-org/skypilot/blob/b147c9ac755af7351b31d91ed5c7520bba2d07f4/sky/spot/controller.py#L149-L154

@Michaelvll @concretevitamin what do you think?

**Comment by concretevitamin** [comment]
Thanks. Read everything before "In conclusion" and they made sense to me.

Question: from the log pasted:
```
(sky-de99-weichiang, pid=734143) I 12-10 21:39:39 controller.py:118] Cluster is preempted (status: INIT). Recovering...
(sky-de99-weichiang, pid=734143) I 12-10 21:42:11 spot_state.py:134] === Recovering... ===
(sky-de99-weichiang, pid=734143) I 12-10 21:42:11 recovery_strategy.py:126] Ignoring the job cancellation failure; the spot cluster is likely completely stopped.
(sky-de99-weichiang, pid=734143) I 12-10 21:42:11 recovery_strategy.py:126]   Detailed exception: 'NoneType' object has no attribute 'cluster_yaml'
(sky-de99-weichiang, pid=734143) D 12-10 21:42:11 recovery_strategy.py:316] Terminating unhealthy spot cluster.
(sky-de99-weichiang, pid=734143) D 12-10 21:42:11 recovery_strategy.py:320] Relaunch the cluster  without constraining to prior cloud/region.
```
we see that `recovery_strategy.py:316] Terminating unhealthy spot cluster.` This suggests a termination request is being run for the preempted cluster (our status: INIT; console status: READY/PREEMPTED). Is this call not going through or otherwise not sufficient? 


--

RE the new method in Resources: it's probably not enough to determine whether it's restartable by looking at the logical representation sky.Resources. For example, the .j2 template can change preemption behavior to stopped.

**Comment by infwinston** [comment]
> we see that recovery_strategy.py:316] Terminating unhealthy spot cluster. This suggests a termination request is being run for the preempted cluster (our status: INIT; console status: READY/PREEMPTED). Is this call not going through or otherwise not sufficient?

Ah this action was in effect only after this PR. Without this PR, the termination action won't be performed.

> RE the new method that takes in a Resources: it's probably not enough to determine whether it's restartable by looking at the logical representation sky.Resources. For example, the .j2 template can change preemption behavior to stopped.

Yes this is a problem. I was thinking we need to add a new field (such as `self.termination_on_spot`) in `Resources` to indicate such property. And each cloud may have their own default value.

**Comment by Michaelvll** [comment]
Thanks for the update @infwinston!

> Ah this action was in effect only after this PR. Without this PR, the termination action won't be performed.

It seems we skip the "retry in the same region first" behavior. That is because the `terminate_cluster` resets the `self._launched_cloud_region`, before the recovery strategy starts. Let's remove the `terminate_cluster` method in the `FailoverStrategyExecutor` and add the `self._launched_cloud_region = None` before the `self.terminate_cluster()` in the `recover()` method.

> RE the new method that takes in a Resources: it's probably not enough to determine whether it's restartable by looking at the logical representation sky.Resources. For example, the .j2 template can change preemption behavior to stopped.

I think the problem may not be related to the preemption behavior, as the `is_spot_restartable` here only means if we can `sky.launch` a cluster with the same name and hash, when the spot cluster is preempted (no matter it is stopped or terminated).

That said, I would prefer to rename the method `is_spot_restartable` to `need_termination_after_preemption`, or something similar.

**Comment by infwinston** [comment]
> It seems we skip the "retry in the same region first" behavior. That is because the terminate_cluster resets the self._launched_cloud_region, before the recovery strategy starts. Let's remove the terminate_cluster method in the FailoverStrategyExecutor and add the self._launched_cloud_region = None before the self.terminate_cluster() in the recover() method.

Great catch! I just updated the code.

> I think the problem may not be related to the preemption behavior, as the is_spot_restartable here only means if we can sky.launch a cluster with the same name and hash, when the spot cluster is preempted (no matter it is stopped or terminated).
That said, I would prefer to rename the method is_spot_restartable to need_termination_after_preemption, or something similar.

Yes, I agree `need_termination_after_preemption` can be a better name. I update the code with a shorter name `need_cleanup_after_preemption`.

However, I think the issue @concretevitamin mentioned still remains? If in our j2 template we specify a different interruption behavior such as `Stop` rather than `Terminate` (default) in https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/interruption-behavior.html.
Then for AWS Spot VM, we shouldn't clean up the spot vm after preemption as we want to reuse it if possible?

**Comment by Michaelvll** [comment]
> However, I think the issue @concretevitamin mentioned still remains? If in our j2 template we specify a different interruption behavior such as Stop rather than Terminate (default) in https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/interruption-behavior.html.
Then for AWS Spot VM, we shouldn't clean up the spot vm's disk after preemption as we want to reuse it if possible?

For those, we still retry on the same region first right? The current change does not make any worse for supporting those in the future comparing to the master branch, right?

**Comment by infwinston** [comment]
> For those, we still retry on the same region first, which will automatically reuse the stopped cluster if needed right?

Ah I see. so you mean for AWS `need_cleanup_after_preemption` should still be `False` even if we change its preemption behavior.
https://github.com/skypilot-org/skypilot/pull/1500/files#diff-96e0e94ebec2d13f2565051ef5df13f97627aa02ad198916e5591392a73d1b65R362
Because AWS never requires manual cleanup by users after preemption.

I think this makes sense.

**Comment by Michaelvll** [comment]
> Ah I see. so you mean for AWS need_cleanup_after_preemption should still be False even if we change its preemption behavior.
https://github.com/skypilot-org/skypilot/pull/1500/files#diff-96e0e94ebec2d13f2565051ef5df13f97627aa02ad198916e5591392a73d1b65R362
Because AWS never requires manual cleanup by users after preemption.

Yes. Since our `sky status -r ` will reflect the correct status from the cloud provider, I think it should be sufficient to handle the different behavior of terminate/stop.

**Comment by infwinston** [comment]
Launched several long running jobs on Spot TPU VM/Pods a few days ago.
They survived through 30+ preemptions in total and no leaked resource found.
Looks like the change should be robust enough..

```
(sky-tmp) weichiang@mbp sky % sky spot queue
Fetching managed spot job statuses...
Managed spot jobs:
In progress jobs: 7 RUNNING

ID  NAME                RESOURCES         SUBMITTED   TOT. DURATION       JOB DURATION        #RECOVERIES  STATUS
19  sky-c6c2-weichiang  1x [tpu-v2-8:1]   3 days ago  3 days 12h 20m 1s   3 days 9h 53m 32s   10           RUNNING
18  sky-f148-weichiang  1x [tpu-v2-8:1]   3 days ago  3 days 21h 10m 15s  3 days 20h 14m 52s  5            RUNNING
17  sky-05c6-weichiang  1x [tpu-v2-8:1]   3 days ago  3 days 21h 11m 56s  3 days 20h 22m 54s  4            RUNNING
16  sky-de99-weichiang  1x [tpu-v2-8:1]   3 days ago  3 days 21h 13m 3s   3 days 19h 27m 23s  11           RUNNING
15  sky-55ca-weichiang  1x [tpu-v3-8:1]   4 days ago  4 days 10h 30m 56s  4 days 10h 1m 3s    4            RUNNING
14  sky-b46d-weichiang  1x [tpu-v2-32:1]  4 days ago  4 days 11h 11m 21s  4 days 10h 32m 46s  4            RUNNING
13  sky-94d0-weichiang  1x [tpu-v3-32:1]  4 days ago  4 days 11h 13m 34s  4 days 10h 18m 54s  4            RUNNING
```

File: sky/backends/backend_utils.py [hunk 2]
```diff
@@ -1228,34 +1228,67 @@ def _get_tpu_vm_pod_ips(ray_config: Dict[str, Any],
...
                            '**** CMD ****\n'
-                           '{tpuvm_cmd}')
+                           f'{query_cmd}\n')
         with ux_utils.print_exception_no_traceback():
-            raise RuntimeError(
-                failure_massage.format(stdout=stdout,
-                                       stderr=stderr,
-                                       tpuvm_cmd=tpuvm_cmd))
-    all_ips = re.findall(IP_ADDR_REGEX, stdout)
+            raise RuntimeError(failure_massage)
```
Comment (Michaelvll) [inline sky/backends/backend_utils.py:1245]: Why don't we use the `subprocess_utils.handle_returncode`?

**Review by Michaelvll — COMMENTED** [review]
Thank you for the fix @infwinston! It looks good to me. Left some comments to make the code cleaner. It would be great if we can run the smoke tests before merging.

File: sky/backends/backend_utils.py [hunk 3]
```diff
@@ -1228,34 +1228,67 @@ def _get_tpu_vm_pod_ips(ray_config: Dict[str, Any],
...
+                               f'{tpuvm_cmd}\n')
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(failure_massage)
+
+        tpuvm_json = json.loads(stdout)
+        if tpuvm_json['state'] != 'READY':
+            # May be a leaked preempted resource.
+            logger.warning(f'TPU VM {tpu_id} is not in READY state. '
+                           'Could be a garbage resource. Skipping...')
+            continue
```
Comment (Michaelvll) [inline sky/backends/backend_utils.py:1278]: Will this state be different for differnet tpu_id?

File: sky/backends/cloud_vm_ray_backend.py [hunk 4]
```diff
@@ -2684,15 +2685,25 @@ def teardown_no_lock(self,
...
+                        shell=True,
+                        stream_logs=False,
+                        require_outputs=True)
+
+                    # Needs to create a list as GCP does not allow deleting
+                    # multiple TPU VMs at once
+                    for tpu_id in stdout.splitlines():
+                        terminate_cmds.append(
+                            f'gcloud compute tpus tpu-vm delete --zone={zone} '
+                            f'--quiet {tpu_id}')
```
[the remark this answers is not in the document]
Comment (infwinston) [reply sky/backends/cloud_vm_ray_backend.py]: good idea. code updated

File: sky/clouds/aws.py [hunk 5]
```diff
@@ -351,3 +351,12 @@ def accelerator_in_region_or_zone(self,
...
             accelerator, acc_count, region, zone, 'aws')
+
+    def need_cleanup_after_preemption(
+            self, resources: 'resources_lib.Resources') -> bool:
+        """Returns whether a spot resource needs cleanup after preeemption."""
+        # By default, AWS Spot instances are not restartable after preemption.
+        # "Terminate interrupted Spot Instances (this is the default behavior)"
+        # See: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/interruption-behavior.html # pylint: disable=line-too-long
+        del resources  # unused
+        return False
```
[the remark this answers is not in the document]
Comment (infwinston) [reply sky/clouds/aws.py]: Good idea. I remove the implementation in AWS and Azure and return `False` by default. Also docstring is updated.

File: sky/backends/backend_utils.py (hunk 2 above)
Comment (infwinston) [reply sky/backends/backend_utils.py]: oh good point. fixed!

File: sky/backends/backend_utils.py (hunk 3 above)
Comment (infwinston) [reply sky/backends/backend_utils.py:1268]: Each TPU VM or TPU Pod only maps to a single `tpu_id`. So yes, different `tpu_id` can have different states.
But when this multiple `tpu_id` situation happens, it means there is a leaked resource with the same cluster name as the current one. That's why I print the garbage resource message.

Normally there should be only one `tpu_id` returned with the below query command.
```
query_cmd = (f'gcloud compute tpus tpu-vm list --filter='
                 f'\\(labels.ray-cluster-name={cluster_name}\\) '
                 f'--zone={zone} --format=value\\(name\\)')
```

File: sky/backends/backend_utils.py [hunk 6]
```diff
@@ -1228,34 +1228,67 @@ def _get_tpu_vm_pod_ips(ray_config: Dict[str, Any],
...
-                                       stderr=stderr,
-                                       tpuvm_cmd=tpuvm_cmd))
-    all_ips = re.findall(IP_ADDR_REGEX, stdout)
+            raise RuntimeError(failure_massage)
+    if len(stdout) == 0:
+        logger.warning('No TPU VMs found with cluster name '
+                       f'{cluster_name} in zone {zone}.')
+    if len(stdout.splitlines()) > 1:
+        logger.warning('Found more than one TPU VM with cluster name '
+                       f'{cluster_name} in zone {zone}.')
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/backends/backend_utils.py]: Isn't it a normal case for spot VM? I think for a non-TPU cluster, we don't show the warning. I think we will handle the number of IPs not equal to the actual amount in the caller function.

File: sky/backends/cloud_vm_ray_backend.py [hunk 7]
```diff
@@ -2684,9 +2684,21 @@ def teardown_no_lock(self,
...
+                        require_outputs=True)
+
+                    # Needs to create a list as GCP does not allow deleting
+                    # multiple TPU VMs at once
+                    tpu_terminate_cmds = []
+                    for tpu_id in stdout.splitlines():
+                        tpu_terminate_cmds.append(
+                            f'gcloud compute tpus tpu-vm delete --zone={zone} '
+                            f'--quiet {tpu_id}')
+                    terminate_cmd = ' && '.join(tpu_terminate_cmds)
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:2701]: We did not handle the first `returncode` here. How about?
```suggestion
                    returncode, stdout, stderr = log_lib.run_with_log(
                        query_cmd,
                        log_abs_path,
                        shell=True,
                        stream_logs=False,
                        require_outputs=True)

                    # Needs to create a list as GCP does not allow deleting
                    # multiple TPU VMs at once
                    # Skip the termination commands, if the TPU ID query commands fail.
                    tpu_terminate_cmds = [f'([[ "{returncode}" == "0" ]] || exit {returncode})']
                    for tpu_id in stdout.splitlines():
                        tpu_terminate_cmds.append(
                            f'gcloud compute tpus tpu-vm delete --zone={zone} '
                            f'--quiet {tpu_id}')
                    terminate_cmd = ' && '.join(tpu_terminate_cmds)
```

**Review by Michaelvll — APPROVED** [review]
LGTM! Thanks @infwinston!

File: sky/clouds/cloud.py [hunk 8]
```diff
@@ -214,5 +214,16 @@ def accelerator_in_region_or_zone(self,
         """Returns whether the accelerator is valid in the region or zone."""
         raise NotImplementedError
 
+    def need_cleanup_after_preemption(self,
+                                      resource: 'resources.Resources') -> bool:
+        """Returns whether a spot resource needs cleanup after preeemption.
+
+        In most cases, spot resources do not need cleanup after preemption.
+        The only exception by far is GCP's Spot TPU VM. We override this method
+        in gcp.py.
```
Comment (Michaelvll) [inline sky/clouds/cloud.py:223]:
```suggestion
        In most cases, spot resources do not need cleanup after preemption,
        as long as the cluster can be launched with the same name and tag,
        no matter the preemption behavior is to terminate or stop the cluster.
        The only exception by far is GCP's Spot TPU VM. We override this method
        in gcp.py.
```

File: sky/backends/backend_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (infwinston) [reply sky/backends/backend_utils.py]: Ah sorry for the confusion. let me explain again. For Spot VM, it also shouldn't happen that multiple Spot TPU VM having the same `labels.ray-cluster-name`. 

Basically the query command should only return one VM in normal case.
```
query_cmd = (f'gcloud compute tpus tpu-vm list --filter='
                 f'\\(labels.ray-cluster-name={cluster_name}\\) '
                 f'--zone={zone} --format=value\\(name\\)')
```
If there's a leak resource (e.g., controller failed to terminate a preempted spot TPU), then this query command will return two VMs which is an abnormal case.

> Also, is it true that a user can have multiple TPU VM with the same name in a same zone?

note that I was not referring to the "TPU name" shown on the console but `labels.ray-cluster-name`. so yes multiple TPU VM can have same `labels.ray-cluster-name`.

[the remark this answers is not in the document]
Comment (infwinston) [reply sky/backends/backend_utils.py]: I'm fine with changing it to `logger.debug` but I'm also afraid that user will never find out there's a leaked resource unless they manually check the console.

[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/backends/backend_utils.py]: I am confused, then why does the problem not happen for a non TPU VM cluster? What ensures those cluster not leaked?

[the remark this answers is not in the document]
Comment (infwinston) [reply sky/backends/backend_utils.py]: for non TPU VM cluster, as it doesn't require manual cleanup after preemption, resource won't be leaked this way? but I'm not sure if there are other scenarios that could trigger leakage. Also we also mostly rely on `ray up` to handle non-TPU VM clusters

File: sky/backends/cloud_vm_ray_backend.py (hunk 7 above)
Comment (infwinston) [reply sky/backends/cloud_vm_ray_backend.py]: good point. fixed with minor modification.

File: sky/clouds/cloud.py (hunk 8 above)
Comment (infwinston) [reply sky/clouds/cloud.py]: good suggestion. fixed with minor modification.

**Review by Michaelvll — APPROVED** [review]

File: sky/backends/cloud_vm_ray_backend.py [hunk 9]
```diff
@@ -2684,9 +2684,24 @@ def teardown_no_lock(self,
...
+                    # multiple TPU VMs at once.
+                    # Skip the termination commands, if the TPU ID
+                    # query command fails.
+                    tpu_terminate_cmds = [f'exit {returncode}'
+                                         ] if returncode != 0 else []
+                    for tpu_id in stdout.splitlines():
+                        tpu_terminate_cmds.append(
+                            f'gcloud compute tpus tpu-vm delete --zone={zone} '
+                            f'--quiet {tpu_id}')
+                    terminate_cmd = ' && '.join(tpu_terminate_cmds)
```
[the remark this answers is not in the document]
Comment (infwinston) [reply sky/backends/cloud_vm_ray_backend.py:2704]: yes let me do it. you meant this right
```
if returncode != 0:
    terminate_cmd = (f'echo "cmd: {query_cmd}" && '
                     f'echo "{stdout}" && '
                     f'echo "{stderr}" >&2 && '
                     f'exit {returncode}')
else:
    tpu_terminate_cmds = []
    for tpu_id in stdout.splitlines():
        tpu_terminate_cmds.append(
            f'gcloud compute tpus tpu-vm delete --zone={zone} '
            f'--quiet {tpu_id}')
    terminate_cmd = ' && '.join(tpu_terminate_cmds)
```

**Comment by infwinston** [comment]
OK all smoke tests have passed. I just launched 7 TPU VMs to see if they can handle preemptions correctly. will merge tomorrow if everything works well.

**Comment by infwinston** [comment]
They all recovered successfully from preemptions (1 out of 7 was the special case situation). Merging this PR now!
```
ID  NAME                                       RESOURCES         SUBMITTED   TOT. DURATION  JOB DURATION   #RECOVERIES  STATUS
46  sky-c3a0-ubuntu                            1x [tpu-v2-8:1]   1 day ago   1 day 26m 28s  1 day 14m 23s  1            RUNNING
45  sky-64ce-ubuntu                            1x [tpu-v3-8:1]   1 day ago   1 day 27m 31s  1 day 15m 27s  1            RUNNING
44  sky-e5d5-ubuntu                            1x [tpu-v2-8:1]   1 day ago   1 day 28m 46s  1 day 14m 20s  1            RUNNING
43  sky-cc0c-ubuntu                            1x [tpu-v2-8:1]   1 day ago   1 day 29m 36s  1 day 12m 18s  1            RUNNING
42  sky-e540-ubuntu                            1x [tpu-v2-8:1]   1 day ago   1 day 30m 22s  1 day 3m 53s   1            RUNNING
41  sky-0d9f-ubuntu                            1x [tpu-v2-32:1]  1 day ago   1 day 31m 37s  1 day 16m 40s  1            RUNNING
40  sky-c11f-ubuntu                            1x [tpu-v3-32:1]  1 day ago   1 day 32m 41s  1 day 15m 10s  1            RUNNING
```
